### PR TITLE
863 adding swagger annotation support for enum and nested objects

### PR DESCRIPTION
**Pull request by fahernandez**

### CODE REVIEW

**Comment by fahernandez** [comment]
Linked to [https://github.com/grpc-ecosystem/grpc-gateway/issues/863](issue) 836

File: examples/proto/examplepb/echo_service.swagger.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "swagger": "2.0",
   "info": {
-    "title": "Echo Service",
+    "title": "examples/proto/examplepb/echo_service.proto",
```
Comment (achew22) [inline examples/proto/examplepb/echo_service.swagger.json]: This seems like an odd change. Why did this become the file path?

File: protoc-gen-swagger/genswagger/template.go [hunk 2]
```diff
@@ -1214,26 +1221,32 @@ func updateSwaggerDataFromComments(swaggerObject interface{}, comment string) er
 
 	paragraphs := strings.Split(comment, "\n\n")
 
-	// If there is a summary (or summary-equivalent), use the first
+	// If there is a summary (or summary-equivalent) and its empty, use the first
```
Comment (achew22) [inline protoc-gen-swagger/genswagger/template.go]: its => it's

File: examples/proto/examplepb/echo_service.swagger.json (hunk 1 above)
Comment (fahernandez) [reply examples/proto/examplepb/echo_service.swagger.json]: I will remove it but the strange thing is that is part of the autogenerated code so I don't know why it is changed.

Comment (fahernandez) [reply examples/proto/examplepb/echo_service.swagger.json]: Actually, I if remove that line, the build process fails https://circleci.com/gh/grpc-ecosystem/grpc-gateway/1982?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: protoc-gen-swagger/genswagger/template.go [hunk 3]
```diff
@@ -408,31 +408,35 @@ func schemaOfField(f *descriptor.Field, reg *descriptor.Registry, refs refMap) s
...
-			updateSwaggerObjectFromJSONSchema(&ret, j)
-		}
-		return ret
 	}
+
+	if j, err := extractJSONSchemaFromFieldDescriptor(fd); err == nil {
+		updateSwaggerObjectFromJSONSchema(&ret, j)
+	}
+
+	return ret
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/template.go]: Is this change necessary? It changes the way we handle `object` and `array` aggregates for no obvious reason to me.

File: protoc-gen-swagger/genswagger/template.go [hunk 4]
```diff
@@ -1189,7 +1196,7 @@ func applyTemplate(p param) (*swaggerObject, error) {
 //
 // If there is no 'Summary', the same behavior will be attempted on 'Title',
 // but only if the last character is not a period.
-func updateSwaggerDataFromComments(swaggerObject interface{}, comment string) error {
+func updateSwaggerDataFromComments(swaggerObject interface{}, comment string, typeName string) error {
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/template.go]: Instead of passing in the typename and adding if cases, can't we skip calling this function altogether if the summary isn't empty?

File: protoc-gen-swagger/genswagger/template.go (hunk 3 above)
Comment (fahernandez) [reply protoc-gen-swagger/genswagger/template.go]: This change allows add field annotations to objects and arrays, before these changes, all field annotations added to object or arrays were ignored.

File: protoc-gen-swagger/genswagger/template.go (hunk 4 above)
Comment (fahernandez) [reply protoc-gen-swagger/genswagger/template.go]: This methods handle specific behavior about setting title/info/description/summary. If we avoid calling this method, this logic needs to be handled on an upper level. The problem is that by default the package title is set to the file name protoc-gen-swagger/genswagger/template.go:949(I don't know why but I suppose we want to keep this behavior) so by changing the behavior of field options taking precedence over comments the title of all protos without package definition was overwritten by the file name so to handle this specific case, the **if** was added

Comment (johanbrandhorst) [reply protoc-gen-swagger/genswagger/template.go]: I see, thanks. Maybe we can change the parameter from a string to a boolean at least? `IsPackageObject`?

Comment (fahernandez) [reply protoc-gen-swagger/genswagger/template.go]: sure, I will change it

File: protoc-gen-swagger/genswagger/template.go (hunk 2 above)
Comment (fahernandez) [reply protoc-gen-swagger/genswagger/template.go]: I will change it

File: examples/proto/examplepb/echo_service.swagger.json (hunk 1 above)
Comment (fahernandez) [reply examples/proto/examplepb/echo_service.swagger.json]: The title of the documents was kept due to some changes when generating the swagger documentation.

**Comment by achew22** [comment]
@fahernandez, thanks so much for finding a problem and reporting it but then going above and beyond to fix it for everyone! Really really great work. I look forward to doing it again soon.
